docs/architecture/authentication-and-authorization.md

```diff
@@ -0,0 +1,125 @@
+# Authentication and Authorization
+
+## Introduction
+Currently, communication between the Compass and both Runtimes and Applications are not secured. We need to provide some security possibilities.
+We want to secure the Compass using ORY's Hydra and OathKeeper. There would be three ways of authentication:
+ - OAuth 2.0 
+ - Client Certificates (mTLS)
+ - JWT token issued by Dex
+
+ There would be three possibilities for securing the connection. To achieve that, first, we need to integrate Hydra and OathKeeper into the Compass. We also need to implement additional supporting components to make our scenarios valid.
+
+## Architecture
+
+The following diagram represents architecture of the security in Compass:
+
+![](./assets/security-architecture.svg)
+
+### Tenant mapping handler
+
+It is a OathKeeper [hydrator](https://github.com/ory/docs/blob/525608c65694539384b785355d293bc0ad00da27/docs/oathkeeper/pipeline/mutator.md#hydrator) handler responsible for mapping `client_id` to tenant. It will be built into the Director component, as a separate endpoint responsible for retrieving the tenant ID based on database query.
+In order to map client_id from Runtime Agent to tenant, we will query `client_id` from Runtime `agentAuth` JSONB field.
+In order to map client_id from Application to tenant, we need to extend Application model with `auth` JSONB field. Then, we can query `client_id` from Application `auth` JSONB field.
+
+To unify tenant mapping, `client_id` will be not only generated in OAuth 2.0 flow, but also in Certificates flow. We will write it into certificate. In a result, these two flows we will map `client_id` to tenant.
+
+### GraphQL security
+
+The Gateway passes request along with JWT token to Compass GraphQL services, such as Director or Connector. The GraphQL components will have authentication middleware and GraphQL [directives](https://graphql.org/learn/queries/#directives) set up for all GraphQL operations (and some specific type fields, if necessary).
```

Comment:
let assume, that there are 10 applications for a given tenant, how do we want to restrict that given app can update only itself, and cannot updates information for other apps?

Reply:
I've added another section